jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/AnsiLoggerWithFooter.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.common.util.concurrent.Futures;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Keeps all log messages in a sequential, deterministic order along with an additional footer that
+ * always appears below log messages. This is intended to log both the messages and the footer to
+ * the same console.
+ *
+ * <p>Make sure to call {@link #shutDown} when finished.
+ */
+class AnsiLoggerWithFooter {
+
+  /** ANSI escape sequence for moving the cursor up one line. */
+  private static final String CURSOR_UP_SEQUENCE = "\033[1A";
+
+  /** ANSI escape sequence for erasing to end of display. */
+  private static final String ERASE_DISPLAY_BELOW = "\033[0J";
+
+  /** ANSI escape sequence for setting all further characters to bold. */
+  private static final String BOLD = "\033[1m";
+
+  /** ANSI escape sequence for setting all further characters to not bold. */
+  private static final String UNBOLD = "\033[0m";
+
+  private static final Duration EXECUTOR_SHUTDOWN_WAIT = Duration.ofSeconds(1);
+
+  private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+  private final Consumer<String> plainLogger;
+
+  private List<String> footerLines = Collections.emptyList();
+
+  AnsiLoggerWithFooter(Consumer<String> plainLogger) {
+    this.plainLogger = plainLogger;
+  }
+
+  /** Shuts down the {@link #executorService}. */
+  public void shutDown() {
+    executorService.shutdown();
+    try {
+      if (!executorService.awaitTermination(
```

Comment:
I think it's better to have `shutdown()` just call `executorService.shutdown()` and not block. I think in many cases, the client of this class don't want to wait after calling `shutdown()` but just proceed. If you want to give the client an option to be able to wait termination, I'd just create another method named `awaitTermination()`.